### PR TITLE
Status and Cromshell config unit test

### PR DESCRIPTION
**Pull request by bshifaw**
Addition of files to test functions in and associated with the status command

### CODE REVIEW

File: tests/unit/mock_data/failed_workflow_slim_metadata.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            "executionStatus":"Failed",
+            "backendStatus":"Failed",
+            "shardIndex":-1,
+            "callRoot":"gs://broad-dsp-lrma-cromwell/CallSVs/6fdae44e-86f1-45e0-8698-db323a66954b/call-SVIM",
+            "attempt":1
+         }
+      ]
+   },
+   "id":"6fdae44e-86f1-45e0-8698-db323a66954b"
+}
```
Comment (bshifaw) [inline tests/unit/mock_data/failed_workflow_slim_metadata.json:42]:
```suggestion
}

```

File: tests/unit/mock_data/running_workflow_slim_metadata.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            "executionStatus":"Done",
+            "backendStatus":"Success",
+            "shardIndex":0,
+            "callRoot":"gs://broad-dsp-lrma-cromwell/PB10xMasSeqSingleFlowcell/54e6a205-8564-429f-8167-1667611a5b17/call-FinalizeSamStatsOnInputBam/shard-0",
+            "attempt":1
+         }
+      ]
+   },
+   "id":"54e6a205-8564-429f-8167-1667611a5b17"
+}
```
Comment (bshifaw) [inline tests/unit/mock_data/running_workflow_slim_metadata.json:63]:
```suggestion
}

```

File: tests/unit/mock_data/succeeded_workflow_slim_metadata.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+         {
+            "callRoot":"gs://broad-dsp-lrma-cromwell/PBCCSOnlySingleFlowcell/a2d4fd6c-86be-4d73-9d00-a46455c6072a/call-FindBams",
+            "executionStatus":"Done",
+            "attempt":1,
+            "shardIndex":-1
+         }
+      ]
+   },
+   "id":"a2d4fd6c-86be-4d73-9d00-a46455c6072a"
+}
```
Comment (bshifaw) [inline tests/unit/mock_data/succeeded_workflow_slim_metadata.json:71]:
```suggestion
}

```

File: tests/unit/mock_data/all.workflow.database.tsv [hunk 4]
```diff
@@ -0,0 +1,6 @@
+DATE	CROMWELL_SERVER	RUN_ID	WDL_NAME	STATUS	ALIAS
+20200110_105202	https://cromwell-v47.dsde-methods.broadinstitute.org	a63aa10c-a43e-4ca7-9be9-c2d2aa08b96d	ValidateBam.wdl	Failed	wombat
+20200110_110109	https://cromwell-v47.dsde-methods.broadinstitute.org	b3b197b3-fdca-4647-9fd8-bf16d2cb734d	ValidateBam.wdl	Succeeded
+20200110_111429	https://cromwell-v47.dsde-methods.broadinstitute.org	682f3e72-0285-40ec-8128-1feb877706ce	ValidateBam.wdl	Succeeded
+20200507_080200	https://cromwell-v45.dsde-methods.broadinstitute.org	d689adec-c600-4e4b-be37-4e30e65848c7	PBCCSOnlySingleFlowcell.wdl	Succeeded	trouble
+20200508_100413	https://cromwell-v45.dsde-methods.broadinstitute.org	d7cad898-f13a-4547-b301-f0c9dbb8865d	PBCCSOnlySingleFlowcell.wdl	Failed
```
Comment (bshifaw) [inline tests/unit/mock_data/all.workflow.database.tsv:6]:
```suggestion
20200508_100413	https://cromwell-v45.dsde-methods.broadinstitute.org	d7cad898-f13a-4547-b301-f0c9dbb8865d	PBCCSOnlySingleFlowcell.wdl	Failed

```

File: tests/unit/mock_data/doom_workflow_slim_metadata.json [hunk 5]
```diff
@@ -0,0 +1,2503 @@
+{
```
Comment (jonn-smith) [inline tests/unit/mock_data/doom_workflow_slim_metadata.json:1]: Is there a reason we need a file this big?

File: tests/unit/test_cromshellconfig.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    def test_resolve_cromwell_config_server_address(self):
+        reload(cromshellconfig)
+        test_cromwell_url = "https://cromwell-v1.dsde-methods.broadinstitute.org"
+
+        assert cromshellconfig.cromwell_server is not None, \
+            "Cromwell server variable should not be empty"
+
+        cromshellconfig.resolve_cromwell_config_server_address(
+            server_user=test_cromwell_url)
```
Comment (jonn-smith) [inline tests/unit/test_cromshellconfig.py:40]:
```suggestion
            server_user=test_cromwell_url
        )
```

**Review by jonn-smith — CHANGES_REQUESTED** [review]
A couple of minor things

File: tests/unit/mock_data/doom_workflow_slim_metadata.json (hunk 5 above)
Comment (bshifaw) [reply tests/unit/mock_data/doom_workflow_slim_metadata.json:1]: No reason for its need, this was just a real doomed workflow metadata that was available

File: tests/unit/test_status.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import pytest
+
+from cromshell.status import command as status_command
+
+
+class TestStatus:
+    """Test the status command functions"""
+
+    def test_workflow_failed_doom(self, mock_data_path):
+        workflow_metadata_path = os.path.join(
```
Comment (SHuang-Broad) [inline tests/unit/test_status.py:13]: failed and doom are two different statuses to me, so maybe we could name it differently

File: tests/unit/test_status.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        with open(workflow_metadata_path, "r") as f:
+            workflow_metadata = json.load(f)
+
+        assert status_command.workflow_failed(workflow_metadata) is True, (
+            "A running doomed workflow metadata should have "
+            "output 'True' to indicate workflow "
+            "has failed."
+        )
+
+    def test_workflow_failed_running(self, mock_data_path):
```
Comment (SHuang-Broad) [inline tests/unit/test_status.py:25]: ditto

File: tests/unit/test_cromshellconfig.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import pytest
+import os
+
+
+class TestCromshellConfig:
+    """Test the cromshell config functions and variables"""
+
+    def test_override_slim_metadata_parameters(self):
+        reload(cromshellconfig)
+        default_slim_parameters = (
```
Comment (SHuang-Broad) [inline tests/unit/test_cromshellconfig.py:13]: There should be a way to place this in the cromshell_config object and read it from the module directly.
This way we avoid the potential problem of having two default values are sometime later the two go out of sync.

File: tests/unit/test_cromshellconfig.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            cromshellconfig.cromwell_server == mock_cromwell_url
+        ), f"Cromwell server variable should be set to {mock_cromwell_url}"
+
+        # Cromwell url parameter should supersede workflow id look up
+        cromshellconfig.resolve_cromwell_config_server_address(
+            server_user=test_cromwell_url, workflow_id=mock_workflow_id
+        )
+        assert (
+            cromshellconfig.cromwell_server == test_cromwell_url
+        ), f"Cromwell server variable should be set to {test_cromwell_url}"
```
Comment (SHuang-Broad) [inline tests/unit/test_cromshellconfig.py:76]: I think there are several cases that are not tests yet.
E.G. what if a valid ID is given, but it happens to be an ID that is neither in the local database `all.workflow.database.tsv`, nor run with the default server address?

File: tests/unit/test_cromshellconfig.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        ).exists(), "Config directory should exist "
+
+    def test_submission_file(self):
+        reload(cromshellconfig)
+        assert (
+            cromshellconfig.submission_file is not None
+        ), "Submission file variable should be set "
+
+        path_to_submission = os.path.join(
+            Path.home(), ".cromshell", "all.workflow.database.tsv"
```
Comment (SHuang-Broad) [inline tests/unit/test_cromshellconfig.py:97]: How about making `all.workflow.database.tsv` a CONSTANT in the cromshell_config module?

File: tests/unit/test_cromshellconfig.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test__load_cromshell_config_options(self, mock_data_path):
+        reload(cromshellconfig)
+        cc__load_cromshell_config_file = getattr(
+            cromshellconfig, "__load_cromshell_config_file"
+        )
+        returned_value = cc__load_cromshell_config_file(os.path.dirname(mock_data_path))
+
+        assert returned_value is not None, "Variable should not be empty"
+        assert type(returned_value) is dict, "Variable should be a dictionary"
```
Comment (SHuang-Broad) [inline tests/unit/test_cromshellconfig.py:121]: I think this test should be expanded:
currently the mock config file has only one line setting the cromwell server, but what about we add more config settings in that file?

Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:121]: Add Todo for this

File: tests/unit/test_status.py (hunk 7 above)
Comment (bshifaw) [reply tests/unit/test_status.py:13]: Be more descriptive with the title of the function, example test_workflow_failed_metadata_doom. Try looking at pythonic suggestion or look at gatk test functions.

File: tests/unit/test_cromshellconfig.py (hunk 9 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:13]: After taking a closer look, line 18 pulls in the cromshell_config object directly and this assert statement is to check whether there's been any changes to the default. If I did make the change to pull in the variable value from the module directly in this line then line 18 would just be checking two of the same variables. I could remove this test function if its not useful?

File: tests/unit/test_cromshellconfig.py (hunk 11 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:97]: done

File: tests/unit/test_cromshellconfig.py (hunk 12 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:121]: done

File: tests/unit/test_status.py (hunk 7 above)
Comment (bshifaw) [reply tests/unit/test_status.py:13]: done, 
A quick search shows the common convention is to  "start with test + name of the code/unit separated by an underscore"

File: tests/unit/test_status.py (hunk 8 above)
Comment (bshifaw) [reply tests/unit/test_status.py:25]: done

File: tests/unit/test_cromshellconfig.py (hunk 9 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:13]: or maybe just check that its not empty

File: tests/unit/test_cromshellconfig.py (hunk 10 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:76]: Added the following tests
1. test when both are given 
2. test when given workflow id that isn’t found in submission file

Also divided the tests up into different functions with clear titles

**Comment by bshifaw** [comment]
@SHuang-Broad ready for rereview

File: tests/unit/test_cromshellconfig.py (hunk 9 above)
Comment (SHuang-Broad) [reply tests/unit/test_cromshellconfig.py:22]: Yeah, I think a null/empty check should be good enough.

File: tests/unit/test_cromshellconfig.py (hunk 10 above)
Comment (SHuang-Broad) [reply tests/unit/test_cromshellconfig.py:115]: 👍

File: src/cromshell/utilities/cromshellconfig.py [hunk 13]
```diff
@@ -22,7 +22,8 @@
 cromwell_api = None
 # Defaults for variables will be set after functions have been defined
 config_dir = None
-submission_file = None
+submission_file_name = "all.workflow.database.tsv"
```
Comment (SHuang-Broad) [inline src/cromshell/utilities/cromshellconfig.py:25]: Unsure if it's a pythonic convention, but typically for constants/macros, let's use uppercase letters to indicate it's a constant.

BTW, is it something intended to be configurable?

Comment (SHuang-Broad) [inline src/cromshell/utilities/cromshellconfig.py:25]: Unsure if it's a pythonic convention, but typically for constants/macros, let's use uppercase letters to indicate it's a constant.

BTW, is it something intended to be configurable?

File: tests/unit/test_cromshellconfig.py (hunk 12 above)
Comment (SHuang-Broad) [reply tests/unit/test_cromshellconfig.py:160]: as a Github issue or as a TODO in the code?

**Review by SHuang-Broad — APPROVED** [review]
Minor comments.

File: tests/unit/test_cromshellconfig.py (hunk 12 above)
Comment (bshifaw) [reply tests/unit/test_cromshellconfig.py:160]: In the code, Line 126 of cromshellconfig.py

File: src/cromshell/utilities/cromshellconfig.py (hunk 13 above)
Comment (bshifaw) [reply src/cromshell/utilities/cromshellconfig.py:25]: Done, Also did the same for the cromshell config file

[PEP](https://www.python.org/dev/peps/pep-0008/#id48)
> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include MAX_OVERFLOW and TOTAL.

**Pull request merged**